### PR TITLE
Adding embed button

### PR DESCRIPTION
**Pull request by vijaynandwani**
To use the embed button:

1) Go to Configuration -> Content authoring ->Text formats and editors and enable the button. 
2) Check "Display embedded entities" and uncheck "Restrict images to this site"

Few questions:

1) Currently I am using the same icon which is being used by File browser module. I have found few other MIT licensed icons. Should we use this one or some other icon?
2) I had to create a new entity browser because the embed button worked only with iframe display plugin. The media library EB uses the modal display plugin. With modal display plugin it gave an error:
"Fatal error: Call to a member function id() on a non-object in /var/www/html/drupaltest/modules/entity_embed/src/Form/EntityEmbedDialog.php on line 524"
3) Because of some reason, the Inline entity forms do not have a submit button. I couldn't find the exact reason behind this problem
4) I am working on tests for the embed button and will update it soon.

### CODE REVIEW

**Comment by paranojik** [comment]
Well... first of all: 1 and 2 should be done automatically, or am I missing something?

**Comment by slashrsm** [comment]
Unfortunately not :( There is no easy way to do that for existing text formats. The only way would be to have an install hook that updates them. However, I am not sure if we want this module to be that opinionated. Maybe we could ship with a format that has this enabled by default? Would that make any sense?

**Comment by mortenson** [comment]
I've wondered about this problem with Entity Embed/File Browser, and was wondering if it makes sense to add a checkbox to all text format configuration forms that just says "Enable Entity Embed Support", then if that is checked modify the filters in an alter hook. This is all spitballing as I haven't checked to see how this would be implemented yet.

**Comment by paranojik** [comment]
IMO we have to make "media" dumb proof. We've made that assumption when we decided to create a number bundles few people will ever use (and configuring a bundle is easy). Why then not facilitate this part which is not straightforward at all?

**Comment by slashrsm** [comment]
The idea that @mortenson proposed is interesting. I am not sure if that should live in entity_embed or media. First one seems a better candidate to me. 

I agree that we should try to make this as simple as possible. Maybe we can try to implement alter hook approach here. We can check a settings variable so people can opt-out if they don't want this. The if we decide to add a UI part we can re-use the same code just check different conditions before executing it. How does that sound?

**Comment by paranojik** [comment]
Yup, I agree. First solve it and then work backwards (by adding UI and stuff). Probably same thing we should do with the bundles users might not need.

File: media.install [hunk 1]
```diff
@@ -62,4 +63,13 @@ function media_install() {
   $source = drupal_get_path('module', 'media') . '/images/icons';
   $destination = \Drupal::config('media_entity.settings')->get('icon_base');
   media_entity_copy_icons($source, $destination);
+  // Add an icon for Media if we aren't installing from configuration.
+  if (!\Drupal::isConfigSyncing()) {
+    $data = file_get_contents(dirname(__FILE__) . '/images/media_embed_icon.png');
+    $file = file_save_data($data, 'public://media_embed_icon.png', FILE_EXISTS_REPLACE);
+    // Set file uuid same as default config.
+    $uuid = (new ExtensionInstallStorage(\Drupal::service('config.storage')))->read('embed.button.media')['icon_uuid'];
```
Comment (slashrsm) [inline media.install:75]: This is very strange. We should rather update the embed button with the UUID of the new file.

Comment (mortenson) [inline media.install:75]: This fix is likely pulled from file_browser.install, you can read through the discussion here: https://www.drupal.org/node/2696593

Comment (slashrsm) [inline media.install:75]: Thank you for the heads-up! In that case we need to add a comment to explain the reason for the non-obvious implementation.

**Comment by paranojik** [comment]
@isthegeek I must say you produce the most informative commit messages :)

**Comment by vijaynandwani** [comment]
@slashrsm @paranojik Tests fail because of this issue: https://www.drupal.org/node/2778843

Will upload the tests for embed button soon.

**Comment by slashrsm** [comment]
That patch committed.

File: media.install [hunk 2]
```diff
@@ -58,8 +60,41 @@ function media_install() {
...
+    if ($editor) {
+      // Make the changes to editor and filters only if editor type exists.
+      $editor_settings = $editor->getSettings();
+      $editor_settings['toolbar']['rows'][0][3]['items'][] = 'media';
+      $editor->setSettings($editor_settings);
+      $editor->save();
+      $format = $editor->getFilterFormat();
+      $format->filters('filter_html')->settings['allowed_html'] .= '<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-settings data-align data-caption data-embed-button>';
+      $format->setFilterConfig('entity_embed', ['status' => 1]);
+      $format->setFilterConfig('filter_html_image_secure', ['status' => 0]);
```
Comment (slashrsm) [inline media.install:95]: Why this?

File: media.install [hunk 3]
```diff
@@ -58,8 +60,41 @@ function media_install() {
...
+  foreach ($editor_names as $editor_name) {
+    $editor = Editor::load($editor_name);
+    if ($editor) {
+      // Make the changes to editor and filters only if editor type exists.
+      $editor_settings = $editor->getSettings();
+      $editor_settings['toolbar']['rows'][0][3]['items'][] = 'media';
+      $editor->setSettings($editor_settings);
+      $editor->save();
+      $format = $editor->getFilterFormat();
+      $format->filters('filter_html')->settings['allowed_html'] .= '<drupal-entity data-entity-type data-entity-uuid data-entity-embed-display data-entity-embed-settings data-align data-caption data-embed-button>';
```
Comment (slashrsm) [inline media.install]: We should check if format uses this filter. AFAIK full_html does not.

File: media.install [hunk 4]
```diff
@@ -58,8 +60,41 @@ function media_install() {
...
+    // Set file uuid same as default config.
+    $uuid = (new ExtensionInstallStorage(\Drupal::service('config.storage')))->read('embed.button.media')['icon_uuid'];
+    $file->set('uuid', $uuid);
+    $file->save();
+  }
+
+  // Enable the media embed button and modify filters.
+  $editor_names = [
+    'basic_html',
+    'full_html',
```
Comment (slashrsm) [inline media.install]: Should we enable button for all formats with WYSIWYG editor on the system?

File: tests/src/FunctionalJavascript/EmbedButtonTest.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  /**
+   * Tests that the entity embed dialog is working.
+   */
+  public function testMediaEmbedDialog() {
+    // Find the button and click it to see if the modal opens.
+    $this->drupalGet('node/add/page');
+    $this->find('.cke_button__media')->click();
+    $this->wait();
+    $this->assertSession()->pageTextContains('Select media to embed');
+
```
Comment (slashrsm) [inline tests/src/FunctionalJavascript/EmbedButtonTest.php:58]: Extra newline.

File: media.install (hunk 3 above)
Comment (slashrsm) [inline media.install]: s/data-entity-embed-settings/data-entity-embed-display-settings

**Comment by slashrsm** [comment]
Tests are failing.

**Comment by vijaynandwani** [comment]
@slashrsm @paranojik Made all the changes.

File: config/install/embed.button.media.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  module:
+    - entity_embed
+    - media_entity
+label: Media
+id: media
+type_id: entity
+type_settings:
+  entity_type: media
+  bundles: {  }
+  display_plugins: {  }
```
Comment (slashrsm) [inline config/install/embed.button.media.yml:15]: I'd limit available plugins here. Only thing that makes sense in my opinion is label and the one that renders entity in dedicated "Embed" view mode. 

We should create that view mode and configure it to have decent look and feel with all bundles.

**Comment by slashrsm** [comment]
Discussed https://github.com/drupal-media/media/pull/15#discussion_r74282934 with @paranojik and he suggested to do this as a follow-up.

**Pull request merged**